guides/initialization.md

```diff
@@ -0,0 +1,131 @@
+
+If a notification `workspace/didChangeConfiguration` or the delayed message is handled, it updates the server settings and triggers builders/analyzers and so on.
+
+## Starts building/analyzing the project
+
+The trigger for the first build is handled on the function `set_settings/2` of the server. After some initialization of mixEnv and other settings, it calls `trigger_build/1`.
+
+The build process itself is handled by `ElixirLS.LanguageServer.Build` module. It does that on a separate process and so the function only returns a reference.
+
+This is where it will load the `mix.exs` project file, check if it needs to update/download dependencies and compiles plus analyzes the project.
+
+Here we have the following steps:
+
+- Loads cached deps from Mix. This is needed to compare if it needs to update/download things;
+- After keeping a reference to old cached deps, it clears deps;
+- Then reloads the project. A great many deal of things happen here as we will see in more detail later;
+- With all configuration set, it fetches dependencies if needed;
+- Compiles all sources and retrieve the diagnostics;
+- Sends the result to the server process so that it can keep current result.
+
+### Reloading the project
+
+The core of the language server is having all the code in a project loaded on a running BEAM instance. Many of its functionality needs to call the `:code` APIs or other introspection modules. Many of its server capabilities use [Elixir Sense](https://github.com/elixir-lsp/elixir_sense) to reply back to the client. This library also needs modules to be loaded in a running BEAM instance so that it can call the same APIs.
+
+Loaded modules are managed in the BEAM by the [code server](http://erlang.org/doc/man/code.html). It is this service that provides things like hot-code reloading. In the code server, every module might have two versions: current code and old code. When doing a hot-reload, for example, it loads the new code, tags the current as old code, purges the old old code and then sends a message to all running processes about the code change. Some important notes on hot-code loading can be read [here](https://learnyousomeerlang.com/designing-a-concurrent-application#hot-code-loving).
+
+When the language server needs to "reload" the project it must first:
+
+- purge/delete old code;
+- compile only the project's `mix.exs` file. This is to ensure it is properly sourced and that it can use it for reading project metadata;
+- ensure that it does not override the server logger config;
+
+This process is handled on the private function `reload_project/0`. One interesting trick it uses is that it sets a different build path using `Mix.ProjectStack.post_config/1`. This is the point where it will create a `.elixir_ls` directory on your project. 
```

Comment:
```suggestion
This process is handled on the private function `reload_project/0`. One interesting trick it uses is that it sets a different build path using `Mix.ProjectStack.post_config/1` (which is a private API). This is the point where it will create a `.elixir_ls` directory on your project. 
```